Add some API consideration text. (#225)
cfrg/draft-irtf-cfrg-hpke: draft-irtf-cfrg-hpke.md

The `0xFFFF` AEAD ID is reserved for applications which only use the Export
interface; see {{hpke-export}} for more details.

# API Considerations {#api-considerations}

> The high-level HPKE APIs specified in this document are all fallible.

Reviewer: Nit: Are Seal() and Export() necessarily fallible? I guess Seal() is vaguely fallible at counter overflow, or if you pick EXPORT_ONLY. Export() seems infallible from the specification's perspective. (Of course, it may be fallible in the implementation if, say, allocation fails. But that's an implementation-specific detail.)

Author: They're both fallible if the input sizes [are simply enormous](https://cfrg.github.io/draft-irtf-cfrg-hpke/draft-irtf-cfrg-hpke.html#name-input-length-restrictions), so I just bundled them all together here. (But that may not be too different from a `malloc` failure 🤷.) I dunno. This seems not wrong, so I'm inclined to keep it.

Reviewer: Hehehe, fair enough, although I don't think you typically have that much address space. :-)

For example, `Decap()` can fail if the encapsulated key `enc` is invalid,
and `Open()` may fail if ciphertext decryption fails. This specification
notes cases where explicit errors may occur.

However, implicit errors may also occur. As an example, certain classes of
failures, e.g., malformed recipient public keys, may not yield explicit
errors. For example, for the DHKEM variant described in this specification,
the `Encap()` algorithm may fail when given an invalid recipient public key.
However, other KEM algorithms may not have an efficient algorithm for verifying
the validity of public keys. As a result, an equivalent error may not manifest

> until AEAD decryption at the recipient. As another example, the `AuthDecap()`

Reviewer 2: ```suggestion
until AEAD decryption at the recipient. As another example, DHKEM's `AuthDecap()`
```
(clarify that we are still talking about DHKEM, as some future KEMs might be able to detect wrong sender public keys explicitly)

The `0xFFFF` AEAD ID is reserved for applications which only use the Export
interface; see {{hpke-export}} for more details.

# API Considerations {#api-considerations}

The high-level HPKE APIs specified in this document are all fallible.
For example, `Decap()` can fail if the encapsulated key `enc` is invalid,
and `Open()` may fail if ciphertext decryption fails. The explicit errors
generated throughout this specification, along with the conditions that
lead to each error, are as follows:

- `ValidationError`: Public key validation failure; {{dhkem}}.
- `DeserializeError`: Public or private key deserialization failure; {{base-crypto}}.
- `DecapError`: `Decap()` failure; {{base-crypto}}.
- `OpenError`: Context AEAD `Open()` failure; {{base-crypto}} and {{hpke-dem}}.
- `MessageLimitReachedError`: Context AEAD sequence number overflow; {{base-crypto}} and {{hpke-dem}}.

> Implicit errors may also occur. As an example, certain classes of failures,

Reviewer 3: I would emphasize that this is not an exhaustive list.  For example, an implementation might simply run out of memory.

e.g., malformed recipient public keys, may not yield explicit errors.
For example, for the DHKEM variant described in this specification,
the `Encap()` algorithm may fail when given an invalid recipient public key.
However, other KEM algorithms may not have an efficient algorithm for verifying
the validity of public keys. As a result, an equivalent error may not manifest
until AEAD decryption at the recipient. As another example, DHKEM's `AuthDecap()`
function will produce invalid output if given the wrong sender public key.
This error is not detectable until subsequent AEAD decryption.

Applications using HPKE APIs should not assume that classes of failures,
especially those that stem from implicit errors, will always manifest the

> same way. Applications SHOULD account for errors when encountered.

Reviewer 3: This is super vague.  What are you actual actually trying to say here?

Author: As in the PR description, I'm trying to say "don't expect only certain types of errors in certain types of places -- be prepared to handle errors anywhere." It would be a shame if an API couldn't adopt a new ciphersuite because their error handling doesn't permit it, for example.

The `0xFFFF` AEAD ID is reserved for applications which only use the Export
interface; see {{hpke-export}} for more details.

# API Considerations {#api-considerations}

The high-level HPKE APIs specified in this document are all fallible.
For example, `Decap()` can fail if the encapsulated key `enc` is invalid,
and `Open()` may fail if ciphertext decryption fails. The explicit errors
generated throughout this specification, along with the conditions that

> lead to each error, are as follows:

Reviewer 4: `VerifyPSKInputs` raises exceptions instead of errors. I'm not sure if that's on purpose. They should probably become errors as well (maybe `ValidationError`?) and get mentioned here.

Author: I think those are best kept as exceptions since they're due to programmer error.

> 

Reviewer 4: The `DeriveKeyPairError` is missing.

Author: Good catch! I'll add it.

- `ValidationError`: KEM input or output validation failure; {{dhkem}}.
- `DeserializeError`: Public or private key deserialization failure; {{base-crypto}}.
- `DecapError`: `Decap()` failure; {{base-crypto}}.
- `OpenError`: Context AEAD `Open()` failure; {{base-crypto}} and {{hpke-dem}}.

> - `MessageLimitReachedError`: Context AEAD sequence number overflow; {{base-crypto}} and {{hpke-dem}}.

Author: ```suggestion
- `MessageLimitReachedError`: Context AEAD sequence number overflow; {{base-crypto}} and {{hpke-dem}}.
- `DeriveKeyPairError`: Key pair derivation failure; {{derivekeypair}}.
```

Implicit errors may also occur. As an example, certain classes of failures,
e.g., malformed recipient public keys, may not yield explicit errors.
For example, for the DHKEM variant described in this specification,

> the `Encap()` algorithm may fail when given an invalid recipient public key.

Reviewer 4: Sec 7.1.4 actually says "the sender MUST validate the recipient's public key pkR"

```suggestion
the `Encap()` algorithm must fail when given an invalid recipient public key.
```

> the `Encap()` algorithm may fail when given an invalid recipient public key.

Author: ```suggestion
the `Encap()` algorithm fails when given an invalid recipient public key.
```

The `0xFFFF` AEAD ID is reserved for applications which only use the Export
interface; see {{hpke-export}} for more details.

# API Considerations {#api-considerations}

The high-level HPKE APIs specified in this document are all fallible.
For example, `Decap()` can fail if the encapsulated key `enc` is invalid,
and `Open()` may fail if ciphertext decryption fails. The explicit errors
generated throughout this specification, along with the conditions that
lead to each error, are as follows:

- `ValidationError`: KEM input or output validation failure; {{dhkem}}.
- `DeserializeError`: Public or private key deserialization failure; {{base-crypto}}.
- `DecapError`: `Decap()` failure; {{base-crypto}}.
- `OpenError`: Context AEAD `Open()` failure; {{base-crypto}} and {{hpke-dem}}.
- `MessageLimitReachedError`: Context AEAD sequence number overflow; {{base-crypto}} and {{hpke-dem}}.

> - `DeriveKeyPairError`: Key pair derivation failure; {{derivekeypair}}.

Author: ```suggestion
- `DeriveKeyPairError`: Key pair derivation failure; {{derive-key-pair}}.
```

| Auth    | y            | y           | y            |
| AuthPSK | y            | y           | y            |

> This analysis assumes that the KEM has no implicit errors.

Reviewer: (I may just be misreading this. I'm not entirely clear on which errors are considered explicit vs implicit here.) It sounds like this sentence, with the rest of the paragraph, is saying that the analysis only holds for DH-based KEMs because non-DH-based KEMs may have implicit errors. But the text above suggests that DH-based KEMs *do* have implicit errors.

Author: It's... tricky :-) I'll leave it to @blipp to clarify precisely what is true about the analysis.

Reviewer 2: This preliminary analysis is only about HPKE with DHKEM. DHKEM is “perfectly correct”, i.e. a valid encapsulation will be successfully decapsulated with probability 1. So, in some sense the analysis indeed assumes that there is no implicit error, but also it is not really an assumption, because we know that DH works that way.

I'd suggest that we do not add this sentence here, but instead add a sentence two paragraphs further down, in the paragraph that discusses the composition theorems proved in the Eurocrypt paper:

> Further, [ABHKLR20] proves composition theorems, showing that HPKE's Auth mode fulfills the security notions of authenticated public key encryption for all KDFs and AEAD schemes specified in this document, given any authenticated KEM satisfying the previously defined security notions for authenticated KEMs. **The theorems assume that the KEM is perfectly correct; it could easily be adapted to work with KEMs that have a non-zero negligible probability for decryption failure.** The assumptions on the KDF are that Extract() and Expand() can be modeled as pseudorandom functions wherein the first argument is the key, respectively. The assumption for the AEAD is IND-CPA and IND-CTXT security.

I will see if I can actually already adapt the composition proofs in CryptoVerif that way and will report back, but I think this is just a technical detail. It boils down to adding one game hop that turns the non-perfectly-correct KEM into a perfectly correct KEM, and accounting for the (negligible) probability that the adversary can detect this change.

On a related note: both analyses do _not_ explicitly model the all-zero/point-at-infinity check of the DH result. However, the theorems are valid for _both_ implementations that do or do not contain this check. (the reason is that it is predictable for the adversary if the DH result will be all-zero, and therefore the adversary could do this check itself before querying the KEM oracle)

Auth mode fulfills the security notions of authenticated public key encryption
for all KDFs and AEAD schemes specified in this document, given any
authenticated KEM satisfying the previously defined security notions
for authenticated KEMs. The theorems assume that the KEM is perfectly correct;

> it could easily be adapted to work with KEMs that have a non-zero but negligible

Reviewer 2: ```suggestion
they could easily be adapted to work with KEMs that have a non-zero but negligible
```
